### PR TITLE
Use regional STS endpoints for IAM join method

### LINKED ISSUE #10415: IAM node joining in China (with auth cluster in aws us-east-2)
### Description

We're operating a single teleport cluster in IoT mode with ~10k nodes connected. We're looking to implement node joining using [these docs here](https://goteleport.com/docs/setup/guides/joining-nodes-aws/), but we have nodes that register that operate in `cn-north-1` region. [Based on this here](https://github.com/gravitational/teleport/blob/4c1695d2e0e87eb3b78ac021863be29f3caa3f04/lib/auth/join_iam.go#L55), it looks like STS won't work because the URL is hardcoded. This issue was opened as a continuation [from this slack thread with](https://gravitational.slack.com/archives/CAHQS4HB5/p1645040410694699) @benarent . 

We haven't reproduced this (yet), so I don't have debug logs or specifics

### Server Details
- Teleport version: 8.3
- Server OS: ubuntu bionic
- Where are you running Teleport? aws
- Additional details: single auth and proxy clusters in aws us-east-2 running in iot mode with nodes connecting from us-east-1, us-east-2, eu-west-1, ap-northeast-1, ap-northeast-2, and cn-north-n1

### PR DESCRIPTION
**Pull request by nklaassen**
Fixes https://github.com/gravitational/teleport/issues/10415

Currently the IAM join method always uses the global STS endpoint `sts.amazonaws.com` as the target for `sts:GetCallerIdentity` requests. These requests are generated and signed on the Node which is joining the cluster, and the Auth server sends them to the actual endpoint to get the response and verify the identity of the Node.

The problem with always using the "global" endpoint, is that it knows nothing about IAM identities in China or GovCloud AWS partitions. Those will need to use the regional endpoint, such as `sts.cn-northwest-1.amazonaws.com.cn`. There are regional endpoints in the default partition as well, we just haven't been using those.

With this change, the joining Node will attempt to find its local region from its environment or the EC2 IMDS, and generate the signed `sts:GetCallerIdentity` request for that regional endpoint.

The Auth server must now send the request to the regional endpoint selected by the client. We need to be very careful that the client cannot trick us into using an attacker-controlled endpoint to attempt to verify the client's identity. To do this, we make sure that the given endpoint is a known valid public STS endpoint we will maintain as a static list.

### Backward Compatibility

The Auth server will always accept join request from Nodes on older versions which use the global STS endpoint, it is still supported.

Upgraded Nodes connecting to an older Auth server (if we release this in a minor, likely 10.2) will still work: they will fall back to using the global endpoint if there is any error trying to use the regional endpoint, which they would hit if the Auth server is on a version which does not yet support regional endpoints.

### CODE REVIEW

**Review by ptgott — APPROVED** [review]

**Comment by nklaassen** [comment]
friendly ping if you guys can review when you get a chance @gabrielcorado and @zmb3 it looks like the bot chose you for code review as well as docs review

File: lib/auth/join_iam.go [hunk 1]
```diff
@@ -49,16 +53,103 @@ const (
...
+//
+// Probably the most secure approach would be to check the given STS URL against
+// a static list of known STS endpoints. This would be hard to maintain over
+// time as AWS adds new regions and partitions.
+//
+// Another option would be to query AWS at runtime to get the list of valid STS
+// endpoints. I couldn't find a suitable API to support this.
+//
+// The approach I've chosen here basically offloads the task of maintaining the
+// static list of STS endpoints to the AWS SDK. Since I can't get at the list
```
Comment (zmb3) [inline lib/auth/join_iam.go:88]: Any concerns about a supply chain attack where someone tricks us into using a modified  version of the AWS SDK and thus using an untrusted endpoint?

I think I'd be okay with it requiring a patch release to support new regions and partitions if it's easier to secure with a hard-coded list.

I believe what you have implemented here to be secure. But there's a fair amount of logic and a lot of special cases to handle, which makes me nervous that we may slip up in the future.

File: lib/auth/join_iam.go [hunk 2]
```diff
@@ -335,16 +426,13 @@ func (a *Server) RegisterUsingIAMMethod(ctx context.Context, challengeResponse c
 
 // createSignedStsIdentityRequest is called on the client side and returns an
 // sts:GetCallerIdentity request signed with the local AWS credentials
-func createSignedStsIdentityRequest(challenge string) ([]byte, error) {
-	// use the aws sdk to generate the request
-	sess, err := session.NewSessionWithOptions(session.Options{
-		SharedConfigState: session.SharedConfigEnable,
-	})
+func createSignedStsIdentityRequest(ctx context.Context, endpointOption stsEndpointOption, challenge string) ([]byte, error) {
```
Comment (zmb3) [inline lib/auth/join_iam.go:429]: While we're here.
```suggestion
func createSignedSTSIdentityRequest(ctx context.Context, endpointOption stsEndpointOption, challenge string) ([]byte, error) {
```

File: lib/auth/join_iam.go [hunk 3]
```diff
@@ -360,3 +448,90 @@ func createSignedStsIdentityRequest(challenge string) ([]byte, error) {
...
+	// use the global STS endpoint, which will not be supported if the AWS identity
+	// being used is for a non-default AWS partition (such as China or
+	// GovCloud.) This is the default behavior on EC2, so let's try to find the
+	// region from the IMDS.
+	if clientConfig := sess.ClientConfig(sts.ServiceName); clientConfig.Endpoint == globalSTSEndpoint {
+		region, err := getEC2LocalRegion(ctx)
+		if trace.IsNotFound(err) {
+			// Unfortunately we could not find the region from the IMDS, go with
+			// the default global endpoint and hope it works.
+			log.Info("Unable to find the local AWS region from the environment or IMDSv2. " +
```
Comment (zmb3) [inline lib/auth/join_iam.go:507]: This is a great message. Informative and actionable. Nicely done.

File: lib/auth/join_iam.go (hunk 1 above)
Comment (nklaassen) [reply lib/auth/join_iam.go:88]: The AWS SDK is compiled into the `teleport` binary, if someone can trick us into using a modified SDK I think there would be much easier and worse ways to pwn us.

But you're right, the validation ended up being more finicky than I would like, and a static list wouldn't be that difficult to maintain, I'll switch to that

Comment (nklaassen) [reply lib/auth/join_iam.go:87]: Done in the latest

**Review by gabrielcorado — APPROVED** [review]

File: lib/auth/join_iam.go [hunk 4]
```diff
@@ -47,19 +50,45 @@ const (
...
+	valid := apiutils.SliceContainsStr(validSTSEndpoints, stsHost)
+	if valid {
+		return nil
+	}
+
+	return trace.AccessDenied("IAM join request uses unknown STS host %q. "+
+		"This could mean that the Teleport Node attempting to join the cluster is "+
+		"running in a new AWS region which is unknown to this Teleport auth server. "+
+		"If this is the case, please check if the endpoint is included in the newest "+
+		"list of known valid endpoints at https://github.com/gravitational/teleport/blob/master/lib/auth/valid_sts_endpoints.go "+
```
Comment (zmb3) [inline lib/auth/join_iam.go:84]: Let me get product approval on this error message.

I'm not sure telling people to check the source code on GitHub is user-friendly enough.

Comment (nklaassen) [reply lib/auth/join_iam.go:84]: I can skip this part and just recommend filing an issue. If they're github savvy enough they'll probably find the source on their own if they want to.

Comment (nklaassen) [reply lib/auth/join_iam.go]: @zmb3 I decided to just reword the comment, it no longer links to the source

**Review by zmb3 — APPROVED** [review]

**Pull request merged**